// @clayui/drop-down/lib/Item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _icon = _interopRequireDefault(require("@clayui/icon"));

var _link = _interopRequireDefault(require("@clayui/link"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ClayDropDownItem = /*#__PURE__*/_react.default.forwardRef(function (_ref, ref) {
  var active = _ref.active,
      children = _ref.children,
      className = _ref.className,
      disabled = _ref.disabled,
      href = _ref.href,
      innerRef = _ref.innerRef,
      onClick = _ref.onClick,
      spritemap = _ref.spritemap,
      symbolLeft = _ref.symbolLeft,
      symbolRight = _ref.symbolRight,
      tabIndex = _ref.tabIndex,
      otherProps = _objectWithoutProperties(_ref, ["active", "children", "className", "disabled", "href", "innerRef", "onClick", "spritemap", "symbolLeft", "symbolRight", "tabIndex"]);

  var clickableElement = onClick ? 'button' : 'span';
  var ItemElement = href ? _link.default : clickableElement;
  return /*#__PURE__*/_react.default.createElement("li", {
    "aria-selected": active,
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(ItemElement, _extends({}, otherProps, {
    className: (0, _classnames.default)('dropdown-item', className, {
      active: active,
      disabled: disabled
    }),
    disabled: disabled,
    href: href,
    onClick: onClick,
    ref: innerRef,
    tabIndex: disabled ? -1 : tabIndex
  }), symbolLeft && /*#__PURE__*/_react.default.createElement("span", {
    className: "dropdown-item-indicator-start"
  }, /*#__PURE__*/_react.default.createElement(_icon.default, {
    spritemap: spritemap,
    symbol: symbolLeft
  })), children, symbolRight && /*#__PURE__*/_react.default.createElement("span", {
    className: "dropdown-item-indicator-end"
  }, /*#__PURE__*/_react.default.createElement(_icon.default, {
    spritemap: spritemap,
    symbol: symbolRight
  }))));
});

var _default = ClayDropDownItem;
exports.default = _default;
