// @clayui/date-picker/lib/DayNumber.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _shared = require("@clayui/shared");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _Helpers = require("./Helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */
var ClayDatePickerDayNumber = function ClayDatePickerDayNumber(_ref) {
  var day = _ref.day,
      daySelected = _ref.daySelected,
      disabled = _ref.disabled,
      _onClick = _ref.onClick;
  var date = day.date,
      outside = day.outside;
  var classNames = (0, _classnames.default)('date-picker-date date-picker-calendar-item', {
    active: date.toDateString() === daySelected.toDateString(),
    disabled: outside || disabled
  });
  return /*#__PURE__*/_react.default.createElement("button", {
    "aria-label": (0, _Helpers.formatDate)((0, _Helpers.setDate)(date, {
      hours: 12,
      milliseconds: 0,
      minutes: 0,
      seconds: 0
    }), 'yyyy MM dd'),
    className: classNames,
    disabled: outside,
    onClick: function onClick() {
      return _onClick(date);
    },
    onKeyDown: function onKeyDown(event) {
      // When tabbing and selecting a DayNumber using
      // SPACE key the active state it's not being removed.
      // See https://github.com/liferay/clay/issues/3374 for more details.
      if (event.key === _shared.Keys.Spacebar) {
        event.preventDefault();
      }
    },
    onKeyUp: function onKeyUp(event) {
      if (event.key === _shared.Keys.Spacebar) {
        _onClick(date);
      }
    },
    type: "button"
  }, date.getDate());
};

var _default = ClayDatePickerDayNumber;
exports.default = _default;
